### PR TITLE
RN-403 Clean up data services configuration

### PR DESCRIPTION
**Pull request by IgorNadj**
RN-403

### Background:

https://github.com/beyondessential/tupaia/pull/3783#discussion_r873263514

This is phase 1 of refactoring isDataRegional style config values.

Phase 1: data_element & data_group
Phase 2+: entity based resolution, detectDataServices, dhis_sync_queue, legacy report configs, legacy map overlay configs

### Changes:
Phase 1 changes:
- Migrate data_source config `isDataRegional: bool` -> `dhisInstanceCode: string|null`
- Create `DhisInstance` as a first class model
- Rename @tupaia/utils/dhis.js functions to legacy_*
- Create new classes and functions in Data Broker which resolve which Dhis Instance to use

---

### Manual release steps:
- [ ] Remove X_DHIS_URL variables after this is released, won't be needed anymore
- [ ] Tell people of the changes, including importing DEs https://beyond-essential.slab.com/posts/importing-data-elements-qhv9v55i
- [ ] Might need to keep an eye on release, central-server logs errors after migration, might need a restart after running "UnhandledPromiseRejectionWarning at getDhisApiInstance"

### CODE REVIEW

File: packages/data-broker/src/__mocks__/utils.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/**
```
Comment (IgorNadj) [inline packages/data-broker/src/__mocks__/utils.js:1]: The existence of this file in __mocks__ tells jest to auto mock the utils package. The alternative of doing it in the test file itself doesn't work due to a limitation of jest.

File: packages/data-broker/src/__tests__/services/dhis/DhisService.stubs.js [hunk 2]
```diff
@@ -34,12 +35,17 @@ export const stubDhisApi = ({
...
       ),
-    getServerName: jest.fn().mockReturnValue(SERVER_NAME),
     getResourceTypes: jest.fn().mockReturnValue({ DATA_ELEMENT: 'dataElement' }),
+    serverName,
   });
-  jest.spyOn(GetDhisApiInstance, 'getDhisApiInstance').mockReturnValue(dhisApi);
+};
 
-  return dhisApi;
+export const stubGetDhisApi = mockDhisApi => {
```
Comment (IgorNadj) [inline packages/data-broker/src/__tests__/services/dhis/DhisService.stubs.js:43]: Split this into two functions because we don't always need to stub as well as mock this import.

File: packages/database/src/migrations/20220513014400-AddDhisInstances-modifies-data.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    readonly: false,
+    config: {
+      productionUrl: 'https://dhis2.palauhealth.org',
+      devUrl: 'https://dev-dhis2.palauhealth.org',
+    },
+  },
+];
+
+/**
+ * DataSources and Entities are linked to a specific DHIS server in many ways:
```
Comment (IgorNadj) [inline packages/database/src/migrations/20220513014400-AddDhisInstances-modifies-data.js:59]: This is the main explainer for how things work... FYI

File: packages/database/src/migrations/20220513014400-AddDhisInstances-modifies-data.js [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  for (const instance of DHIS_INSTANCES) {
+    await insertObject(db, 'dhis_instance', instance);
+  }
+
+  // Regional, we migrate trivially
+  //   - note: undefined defaults to regional
+  await db.runSql(`
+    UPDATE data_source 
+    SET config = (config || '{"dhisInstanceCode": "regional"}') #- '{isDataRegional}'
```
Comment (IgorNadj) [inline packages/database/src/migrations/20220513014400-AddDhisInstances-modifies-data.js:100]: `||` adds properties to jsonb, `#-` removes them.

File: packages/central-server/src/database/utilities/assertCanAddDataElementInGroup.js [hunk 5]
```diff
@@ -3,11 +3,23 @@
...
-const constructErrorMessage = ({ property, value, dataElementCode, otherGroupCode }) =>
-  `Cannot set ${property} to '${value}': data element '${dataElementCode}' is included in data group '${otherGroupCode}', which uses another ${property}`;
+const areDifferentAllowingNull = (a, b) => {
+  if (a === null && b === null) return false;
+  if (a === null && b !== null) return true;
+  if (a !== null && b === null) return true;
+  return a !== b;
+};
+
+const constructErrorMessage = ({
```
Comment (IgorNadj) [inline packages/central-server/src/database/utilities/assertCanAddDataElementInGroup.js:15]: I updated these argument names and the error message for clarity, functionality is the same

**Review by rohan-bes — COMMENTED** [review]
It's looking really good man! Thanks so much for doing this!

Just a few comments, I might give this one a second pass. But first I was hoping we could have a bit of a high level chat? You've added `dhis_instance`s here as a concept, but I had maybe thought we'd be adding `data_services` as a concept, where a data service could be:
1. A specific dhis2 instance
2. A data-lake instance
3. The internal data-api instance
4. etc.

Basically where we currently have data sources attached to a 'service type', they'd instead be linked to a specific service instance, which we could have many of.

Now all that wouldn't necessarily be in scope here, and we'd still need to use a lot of this cleanup code you've written anyway. Either way lets chat about it on Thursday and share thoughts?

File: packages/data-broker/src/services/dhis/getDhisApi.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
```
Comment (rohan-bes) [inline packages/data-broker/src/services/dhis/getDhisApi.js:4]: Love this set of functions being encapsulated in a single file! Really helps make clear the different ways the interface is called!

File: packages/database/src/migrations/20220512232924-CreateDhisInstanceTable-modifies-schema.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  seed = seedLink;
+};
+
+exports.up = async function (db) {
+  await db.runSql(`
+    CREATE TABLE dhis_instance (
+      id TEXT PRIMARY KEY,
+      code TEXT NOT NULL UNIQUE,
+      readonly BOOLEAN NOT NULL,
+      config JSONB NOT NULL
```
Comment (rohan-bes) [inline packages/database/src/migrations/20220512232924-CreateDhisInstanceTable-modifies-schema.js:23]: I'm perfectly happy with raw SQL here, but incase you weren't aware (I wasn't until recently) there's a `db.createTable()` function 👍

File: packages/utils/src/legacyDhis.js [hunk 8]
```diff
@@ -61,22 +63,61 @@ const getServerName = (entityCode, isDataRegional) => {
...
+ * @deprecated Use data-broker instead
+ *
+ * Legacy to modern config mapping function. Clean up after legacy reports are all migrated.
+ *
+ * Note: some parts of Tupaia will not function correctly in deployments where a DHIS Instance
+ * with code 'regional' does not exist
+ *
+ * @param {{ isDataRegional: boolean}} legacyConfig
+ * @return {string|undefined} DhisInstance code
+ */
```
Comment (rohan-bes) [inline packages/utils/src/legacyDhis.js:118]: Thanks for adding this comment!

**Review by rohan-bes — APPROVED** [review]
Nice work @IgorNadj this is a great step forward! Hopefully we can pick up and carry on with phase 2 before too long.

A small comment from the first pass, but overall really happy and approving. Really appreciate the tests and comments btw (this is all becoming rather arcane knowledge)

Thoughts around testing on this? What's the regression risk looking like? Obviously there's a bunch of vizes... Does this break older survey import files too?

File: packages/database/src/migrations/20220513014400-AddDhisInstances-modifies-data.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+ *     - Finds SurveyResponse this answer is linked to, and uses that method with it.
+ * - Pre-aggregation:
+ *   - Always uses regional DHIS, hardcoded
+ *
+ * ## Changes
+ * This migration only tackles a cleanup of the [Standard Way]. This migration introduces DhisInstance as a first
+ * class model. We link data_source to dhis_instance via config option `data_source.config.dhisInstanceCode: <string>`.
+ *
+ * For entity based resolution, where previously we had isDataRegional = false, we now leave dhisInstanceCode
+ * not set, which indicates no specific dhisInstance, and that we need to resolve it from the entity.
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/database/src/migrations/20220513014400-AddDhisInstances-modifies-data.js:86]: Discussed, have created https://linear.app/bes/issue/RN-537/one-data-element-per-dhis-server

File: packages/admin-panel/src/pages/resources/DataSourcesPage.js [hunk 10]
```diff
@@ -77,9 +77,11 @@ const DATA_ELEMENT_FIELDS = [
...
       getJsonFieldSchema: () => [
         {
-          label: 'Regional Server (Choose "No" if stored on country specific server)',
-          fieldName: 'isDataRegional',
-          type: 'boolean',
+          label: 'DHIS Server (Leave blank if stored on country specific server)',
+          fieldName: 'dhisInstanceCode',
+          optionsEndpoint: 'dhisInstances',
+          optionLabelKey: 'dhisInstances.code',
+          optionValueKey: 'dhisInstances.code',
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/admin-panel/src/pages/resources/DataSourcesPage.js:84]: Yeah I don't love it either. Were you thinking something like this? https://linear.app/bes/issue/RN-403#comment-deae01b8

**Comment by IgorNadj** [comment]
TODO: remove ability for users to import/set dhisInstanceCode as null. We don't want new dataElements/groups to use entity based resolution. There are no cases where want to do this out of necessity where a DE/DG lives on multiple DHIS Servers legitimately: https://linear.app/bes/issue/RN-537#comment-bf70ca96

File: packages/data-broker/src/services/dhis/DhisInstanceResolver.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      allDhisInstances.push(dhisInstance);
+    }
+    if (allDhisInstanceCodes.size > 1) {
+      throw new Error(
+        `All entities must use the same DHIS2 instance (got [${[...allDhisInstanceCodes].join(
+          ',',
+        )}])`,
+      );
+    }
+    return allDhisInstances[0];
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/data-broker/src/services/dhis/DhisInstanceResolver.js:74]: I left this as multiple single calls to legacy_getDhisServerName to make mocking easier. This way we can have a test in data-broker to make sure an error gets thrown if it resolves to multiple dhis servers. I thought about having this test in utils, on the legacy_getDhisServerName function, but I thought it might make more sense to have it on the non-legacy side of the codebase rather than the legacy side. I think the reason was that the legacy side is really hard to write tests for because mocking is hard. As this class should disappear once the data is cleaned up, are you happy to leave as is?

File: packages/central-server/src/database/utilities/assertCanAddDataElementInGroup.js [hunk 12]
```diff
@@ -3,11 +3,23 @@
...
+const areBothDefinedAndDifferent = (a, b) => a !== undefined && b !== undefined && a !== b;
 
-const constructErrorMessage = ({ property, value, dataElementCode, otherGroupCode }) =>
-  `Cannot set ${property} to '${value}': data element '${dataElementCode}' is included in data group '${otherGroupCode}', which uses another ${property}`;
+const areDifferentAllowingNull = (a, b) => {
+  if (a === null && b === null) return false;
+  if (a === null && b !== null) return true;
+  if (a !== null && b === null) return true;
+  return a !== b;
+};
```
[the remark this answers is not in the document]
Comment (IgorNadj) [reply packages/central-server/src/database/utilities/assertCanAddDataElementInGroup.js]: Ah good catch, updated.

**Comment by IgorNadj** [comment]
Have rebased on dev, fixed up PR from feedback.

File: packages/data-broker/src/services/dhis/DhisInstanceResolver.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (rohan-bes) [reply packages/data-broker/src/services/dhis/DhisInstanceResolver.js:74]: Yup that sounds reasonable to me 👍

**Review by rohan-bes — APPROVED** [review]
Great stuff @IgorNadj really appreciate all this work!
